### PR TITLE
(WIP) fix(vm.cool) Persist storage changes

### PR DESCRIPTION
**Pull request by ditto-eth**
## Motivation

> See https://github.com/foundry-rs/foundry/pull/5830#discussion_r1326240634 (earlier merged PR) for more context. Last PR was ultimately the wrong approach due to potential issues with side effects and also that `storage.clear()` actually reset the storage so the storage values didn't persist making the feature not useful.

this is a wip, a lot of things to finish up but wanted to get feedback as i finally got somewhere.

- [ ] what needs to be done in relation to refunds/reverts, if at all?
- [ ] removed the account warm/cold for now, just focusing on the slots (@mds1 q was about preventing certain addresses from being `cold` because they are always warm at the start of the tx (precompiles, from, to).
- [ ] test multiple slots/addresses
- [ ] `function cool(StorageSlot[] memory slots) external;` not implemented but should be simple if this approach works (i think this can just be a separate pr)

## Solution

via `revm` telegram:

> mattsse: the issue with gas tracking here is, that gas is always deducted, only refunds are added back
> so even if you clear something, this should not affect gas metering
> I think the only way to do this is playing the `Gas` return value of the Inspector::call{_end} function calls
> rakita: This could potentially introduce sideeffects, as changed storage values gets removed. You can maybe craete your own JournalState (that would need to track reverts) and just override the gas used inside `step_end`

^ approach seems to generally make sense? added a `step_end` to cheatcodes inspector, so on `SSTORE/SLOAD` i can manipulate the `interpreter.gas` value. The idea is that a certain amount of cost needs to be added back depending on `sstore`.

fyi i kept all my `println` calls in case someone wanted to check it out locally, will remove when the pr is ready to go.

```rust
    pub cool: HashMap<B160, bool>,
    // Whether an addresses storage slot is cool or not
    pub cool_storage: HashMap<B160, HashMap<U256, bool>>,
    // How much gas to add to the next step/op code due to a cool storage slot
    pub additional_gas_next_op: u64,
```

`vm.cool` sets the variables ^. It just saves whether an address `B160`'s slots are cool/warm or not by resetting it everytime the `vm.cool` is called. then just calculate the added gas cost (usually +2100), but different if an sstore is done and mark the slot as "warm" so the regular gas is sufficient (until the next `vm.cool` is called if necessary). can add a few more kinds of tests.

> also i'm sure it can be optimized (didn't know any rust till last pr, so understand that there might be simpler refactors

### CODE REVIEW

**Comment by ditto-eth** [comment]
will push some more changes in a bit, fixes some things

**Comment by Evalir** [comment]
just looping in @mds1 for discussing approach — will review soon!

File: crates/evm/src/executor/inspector/cheatcodes/mod.rs [hunk 1]
```diff
@@ -538,6 +558,165 @@ impl<DB: DatabaseExt> Inspector<DB> for Cheatcodes {
...
+        eval: InstructionResult,
+    ) -> InstructionResult {
+        // for cool cheatcode gas costs,
+        // see https://eips.ethereum.org/EIPS/eip-2200, https://eips.ethereum.org/EIPS/eip-2929
+
+        // if previous step added gas, add it once
+        // note that all the opcodes will already have a cost (usually 100)
+        // so adding until it hits the gas expected for a cold key/address
+        if self.cool_additional_gas_next_op > 0 {
+            interpreter.gas.record_cost(self.cool_additional_gas_next_op);
```
Comment (ditto-eth) [inline crates/evm/src/executor/inspector/cheatcodes/mod.rs:575]: main thing that happens is setting the Gas at each op code. since the cost isn't known until the next op code it records it at the next one, and resets. All the other state is just there to track when `vm.cool` is called and each time an opcode is stepped through (when to "consume" the cool state and make it warm) for both the the address and the address' slots

File: crates/evm/src/executor/inspector/cheatcodes/mod.rs [hunk 2]
```diff
@@ -201,6 +202,25 @@ pub struct Cheatcodes {
...
+#[derive(Clone, Debug)]
+pub enum CoolState {
+    CalledAndAccessed,
+    Called,
+}
+
+#[derive(Clone, Debug)]
+pub enum CoolStorageState {
+    Warm,
+    WarmAndSSTORED,
```
Comment (ditto-eth) [inline crates/evm/src/executor/inspector/cheatcodes/mod.rs:223]: don't like these names but just went with it for now

File: crates/evm/src/executor/inspector/cheatcodes/mod.rs [hunk 3]
```diff
@@ -538,6 +558,165 @@ impl<DB: DatabaseExt> Inspector<DB> for Cheatcodes {
...
     }
 
+    #[inline]
+    fn step_end(
+        &mut self,
+        interpreter: &mut Interpreter,
+        data: &mut EVMData<'_, DB>,
+        eval: InstructionResult,
+    ) -> InstructionResult {
+        // for cool cheatcode gas costs,
```
Comment (Evalir) [inline crates/evm/src/executor/inspector/cheatcodes/mod.rs:568]: oof, this adds a ton of code to the main func—can we abstract the `cool` related code to a different function? This way it's easier to change in the future.

**Review by Evalir — COMMENTED** [review]
Still haven't reviewed for correctness—some nits

File: crates/evm/src/executor/inspector/cheatcodes/mod.rs (hunk 3 above)
Comment (ditto-eth) [reply crates/evm/src/executor/inspector/cheatcodes/mod.rs:568]: sounds good, `step_end` wasn't used at all before, but it may look better with one line - i'm not familiar with the codebase's organization so left it as non dry as possible

File: crates/evm/src/executor/inspector/cheatcodes/env.rs [hunk 4]
```diff
@@ -23,7 +23,7 @@ use revm::{
     primitives::{Bytecode, SpecId, KECCAK_EMPTY},
     Database, EVMData,
 };
-use std::collections::BTreeMap;
+use std::collections::{BTreeMap, HashMap};
```
[the remark this answers is not in the document]
Comment (ditto-eth) [reply crates/evm/src/executor/inspector/cheatcodes/env.rs:26]: ok I just did this since there was a `HashMap` in the other file, mod.rs as well, so didn't know which one to use?

[the remark this answers is not in the document]
Comment (ditto-eth) [reply crates/evm/src/executor/inspector/cheatcodes/env.rs:26]: should I import it there too and just alias it to `rHashMap` or something?

File: crates/evm/src/executor/inspector/cheatcodes/env.rs [hunk 5]
```diff
@@ -320,15 +320,12 @@ fn add_breakpoint(state: &mut Cheatcodes, caller: Address, inner: &str, add: boo
...
-// mark the slots of an account and the account address as cold
-fn cool_account<DB: DatabaseExt>(data: &mut EVMData<'_, DB>, address: Address) -> Result {
-    if let Some(account) = data.journaled_state.state.get_mut(&address.to_alloy()) {
-        if account.is_touched() {
-            account.unmark_touch();
-        }
-        account.storage.clear();
-    }
-
+// setup the state to mark the slots of an account and the account address as cold
```
[the remark this answers is not in the document]
Comment (ditto-eth) [reply crates/evm/src/executor/inspector/cheatcodes/env.rs:323]: that's actually pretty funny to me now, even notice the ///s

File: crates/evm/src/executor/inspector/cheatcodes/mod.rs (hunk 3 above)
Comment (Evalir) [reply crates/evm/src/executor/inspector/cheatcodes/mod.rs:568]: yep that's cool!

File: crates/evm/src/executor/inspector/cheatcodes/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mds1) [reply crates/evm/src/executor/inspector/cheatcodes/mod.rs]: FWIW I can't tell what this enum is supposed to represent from the names here alone, so we probably need to change the names and agree with adding docs. `CoolStorageState::Warm` and `CoolStorageState::WarmAndSSTORED` feel like conflicting names, e.g. how is the state cool and warm? Also maybe `WarmWithSSTORE` or `WarmWithSstore` would be easier to read since `SSTORED` isn't a usage I've seen

[the remark this answers is not in the document]
Comment (ditto-eth) [reply crates/evm/src/executor/inspector/cheatcodes/mod.rs]: right, i didn't know what to name the hashmap itself so it's `coolstoragestate` but it's more the storage state that is tracked when calling `vm.cool`, but it's tracking whether it's warm, or warm and the storage store was set, or that it's cool. Maybe very confusing but if it's empty it means it's cool (because each vm.cool resets the whole hashmap to be empty)

so it's either empty value (none), or warm (the sload makes it warm), or warm and sstore was called

I suppose this means cool, sloaded, or sstored, and each of those should change the gas increase

**Comment by ditto-eth** [comment]
ok updated the renames, attempted to fix the ci issues..

**Comment by ditto-eth** [comment]
`test integration (1/3)` keeps failing for me - the local tests pass fine but i don't know how to run these tests locally so I have to guess what the issue is to try to fix it. I assume it's a different build/nightly/foundry toml so the gas has different values depending on the optimizations? it is true these tests are too exact so test could be flaky if solc changes

```sh
thread 'cheats::test_cheats_local' panicked at 'called `Result::unwrap()` on an `Err` value: Test testCool_call() did not pass as expected.
Reason: None
Logs:
Error: a == b not satisfied [uint]
  Expected: 44835
    Actual: 45299
Error: a == b not satisfied [uint]
  Expected: 27735
    Actual: 28199
```

**Comment by Evalir** [comment]
@ditto-eth you can use `cargo t cheats_local`

**Comment by ditto-eth** [comment]
> @ditto-eth you can use `cargo t cheats_local`

ah i'd been using a custom command I made up `cargo r --bin forge test --contracts testdata/cheats/Cool.t.sol --match-test testCool -vv`, but that passes for me so i'm confused by the ci failure (there seems to be a gas difference of `13` in some tests and now others. The command you mentioned passes and either finishes but doesn't exit or is hanging?

File: crates/evm/src/executor/inspector/cheatcodes/mod.rs [hunk 6]
```diff
@@ -201,6 +203,175 @@ pub struct Cheatcodes {
...
+    // note that all the opcodes will already have a cost (usually 100)
+    // so adding until it hits the gas expected for a cold key/address
+    if state.additional_gas_next_op > 0 {
+        interpreter.gas.record_cost(state.additional_gas_next_op);
+        state.additional_gas_next_op = 0;
+    }
+
+    // if cool cheatcode was ever called on this address
+    let contract_address = interpreter.contract().address.to_ethers();
+    if state.addresses.get(&contract_address).is_some() {
```
Comment (ditto-eth) [inline crates/evm/src/executor/inspector/cheatcodes/mod.rs:256]: - `vm.cool` involves two parts: making the `address` itself cool, making the storage slots cool
- check that `vm.cool` was passed to this `address`
  - if so, track a few opcodes (call/balance/etc) to mark the address as warm, as well as queuing up additional gas to be added to the next opcode. a warm address is 100, a cool one is 2600, so add 2500.
- check the storage slots
  - if.an sload or sstore, mark address as warm but add different wants of extra gas depending on the situation.

**Comment by jameswenzel** [comment]
Curious what the behavior will be for cold-marked accounts/slots accessed within reverting callframes ([some context](https://twitter.com/emo_eth/status/1696281639201259677))

For EVM-accurate measurements, reverting callframes will have to "re-cool" `vm.cool`ed slots.

This is something I was wrestling with in my much more roundabout [account/slot enumeration cheatcodes](https://github.com/foundry-rs/foundry/pull/5794).

**Comment by ditto-eth** [comment]
awesome thanks for the feedback/comment!

> Curious what the behavior will be for cold-marked accounts/slots accessed within reverting callframes ([some context](https://twitter.com/emo_eth/status/1696281639201259677))

i'm not that familiar - is this when a new contract/function is called that fails within a try/catch? if so, then yea it would need to deduct the same amount back? i didn't think that would be a common case in the context of a gas test but maybe just ignorant of that (and why I wasn't sure what handling reverts meant as in my mind it would just fail the test anyway)

**Comment by jameswenzel** [comment]
> awesome thanks for the feedback/comment!
> 
> > Curious what the behavior will be for cold-marked accounts/slots accessed within reverting callframes ([some context](https://twitter.com/emo_eth/status/1696281639201259677))
> 
> i'm not that familiar - is this when a new contract/function is called that fails within a try/catch? if so, then yea it would need to deduct the same amount back? i didn't think that would be a common case in the context of a gas test but maybe just ignorant of that (and why I wasn't sure what handling reverts meant as in my mind it would just fail the test anyway)

Yeah, a `try/catch` or a low-level call where `success` is checked/handled/ignored. It's definitely a more niche case – but there are certainly contracts designed to handle failing edge cases gracefully. (A fun one is forcefully transferring native tokens to smart contracts that rejects them, such as [Solady's selfdestruct mechanism](https://github.com/Vectorized/solady/blob/main/src/utils/SafeTransferLib.sol#L87))

It's probably shouldn't be a blocker to shipping the fix, since this helps a ton with getting gas reports more accurate – but full accuracy for complex protocols probably relies on supporting these edge cases.

**Comment by ditto-eth** [comment]
i guess the selfdestruct won't be happening if it's being deprecated later i heard?

> but full accuracy for complex protocols probably relies on supporting these edge cases.

down to make it accurate as possible in the next PRs, as i'm trying to test it on my protocol now which spurred my interest in figuring this out! 

also in writing the tests, i started getting confused what i'd even expect the result to be 🙂 so curious on reviewing that. tests are probably based on number of opt runs which change the opcodes being done (this is annoying to test sload/sstore, as you can't just do a `contract.var` since it will get optimized out, so have to set it to a variable which costs some other amount of gas) - there's no way to just do the basic read/write operations without assembly right? I wonder if one can turn off the compiler settings for specific tests and do sload/sstore in yul, but i'd probably need some help on that?

**Comment by ditto-eth** [comment]
does anyone know how the gas is modified with reverts normally? sounds like i'd need to track the additional gas per frame and then subtract it if revert goes through? maybe the code related to gas metering in `mod.rs` helps, it seems to just reset the gas at various opcodes.

**Comment by jameswenzel** [comment]
> does anyone know how the gas is modified with reverts normally? sounds like i'd need to track the additional gas per frame and then subtract it if revert goes through? maybe the code related to gas metering in `mod.rs` helps, it seems to just reset the gas at various opcodes.

I think moreso that `vm.cool`ed accounts/slots should have the extra gas charged within a callframe – but if that callframe gets reverted, Forge should re-mark the accounts/slots as `vm.cool`ed again. "Cold" accounts/slots will be charged "cold" gas within a reverting callframe, but they will still be cold after the callframe exists and the surcharge will need to be added again if they get accessed again.

(An exception here is reverting CREATEs; the "CREATE"d address always gets marked warm, even if create fails)

**Comment by ditto-eth** [comment]
i'm not sure what else needs to be done for this pr if we move the following points out of this
- moving `function cool(StorageSlot[] memory slots) external;` to be a nice to have (based on the current implementation, one would need to mark all previous state changes as warm first)
- handling reverts/callframe (mark address as cool again, seems like it would be best to re-use the data in https://github.com/foundry-rs/foundry/pull/5794 and simply call `cool_account` on any addresses that need to be re-cooled when it happens)
- not setting the address as `cool` for the `from` and `to` address?

**Comment by mattsse** [comment]
@mds1 I'm a bit out of the loop here

does this sound conceptionally right?

**Comment by ditto-eth** [comment]
> @mds1 I'm a bit out of the loop here
> 
> does this sound conceptionally right?

explained a bit here, but i can break it down more: https://github.com/foundry-rs/foundry/pull/5852#pullrequestreview-1642807423

### `vm.cool` cools both the `address` itself and the address' storage slots.
- this approach is doing a lot work just to modify the gas amounts per opcode to change gasleft, doesn't modify anything else.
- track address state via a new `state.addresses` hashmap
- track storage slots via a double hashmap `state.address_storage` which is by contract address and then by slot number.
- add another state variable to track how much gas to add to the next opcode
- in the cheatcodes/mod.rs, modify the inspector so that at the end of a step, check which opcodes were run. 
  - If they match a sload/sstore (which marks storage as warm, then update the tracked state accordingly and also add the extra gas to the next opcode). there are some special amounts based on the storage values.
  - if they match an address opcode (create/call/balance etc) then mark the address as warm + extra gas.
  
  there are some edge cases I mentioned ^. The only thing i might want to do still is deciding (or documenting) how to handle warm address for `to` and `from`. also looks like i need to rebase

**Comment by ditto-eth** [comment]
light ping, let me know if there's anything else I should explain or add!

**Comment by DaniPopes** [comment]
Hey @ditto-eth, we recently refactored how cheatcodes are implemented in https://github.com/foundry-rs/foundry/pull/6131. You can read more about it [in the dev docs](https://github.com/foundry-rs/foundry/blob/master/docs/dev/cheatcodes.md). Please let me know if you have any problems rebasing!

**Comment by ditto-eth** [comment]
@DaniPopes thanks for the ping, wasn't too bad as I already did a previous rebase so a bit more familiar now - also decided to squash into 1 commit to make that easier.

File: crates/cheatcodes/src/evm.rs [hunk 7]
```diff
@@ -9,7 +10,7 @@ use foundry_common::fs::read_json_file;
 use foundry_evm_core::backend::DatabaseExt;
 use foundry_utils::types::ToAlloy;
 use revm::{
-    primitives::{Account, Bytecode, SpecId, KECCAK_EMPTY},
+    primitives::{Account, Bytecode, HashMap as rHashMap, SpecId, KECCAK_EMPTY},
```
Comment (DaniPopes) [inline crates/cheatcodes/src/evm.rs:13]: Any reason as to why we're not using standard hashmap?

File: crates/cheatcodes/src/evm.rs [hunk 8]
```diff
@@ -302,11 +303,11 @@ impl Cheatcode for storeCall {
...
-        let Self { target } = self;
-        if let Some(account) = ccx.data.journaled_state.state.get_mut(target) {
-            account.unmark_touch();
-            account.storage.clear();
-        }
+        let Self { target } = *self;
+        ensure_not_precompile!(&target, ccx);
+        // TODO: prevent or warn about cooling the to/from address in a tx
+        ccx.state.addresses.insert(target, AddressState::Cool);
+        ccx.state.address_storage.insert(target, rHashMap::new());
```
Comment (DaniPopes) [inline crates/cheatcodes/src/evm.rs:310]: Can these be merged into one map?

File: crates/cheatcodes/src/inspector.rs [hunk 9]
```diff
@@ -200,6 +200,174 @@ pub struct Cheatcodes {
...
+}
+
+/// Function to charge extra gas per opcode based on cool cheatcode
+fn add_gas_from_cool_cheatcode<DB: DatabaseExt>(
+    state: &mut Cheatcodes,
+    interpreter: &mut Interpreter,
+    data: &mut EVMData<'_, DB>,
+) -> InstructionResult {
+    // For gas costs, see https://eips.ethereum.org/EIPS/eip-2200, https://eips.ethereum.org/EIPS/eip-2929
+
```
Comment (DaniPopes) [inline crates/cheatcodes/src/inspector.rs:239]: Can we short circuit this, maybe on checking the maps' length? This function is ran on every opcode so I would like to keep the overhead as little as possible.

File: crates/cheatcodes/src/inspector.rs [hunk 10]
```diff
@@ -200,6 +200,174 @@ pub struct Cheatcodes {
...
+    /// same as if empty
+    Cool,
+}
+
+/// Function to charge extra gas per opcode based on cool cheatcode
+fn add_gas_from_cool_cheatcode<DB: DatabaseExt>(
+    state: &mut Cheatcodes,
+    interpreter: &mut Interpreter,
+    data: &mut EVMData<'_, DB>,
+) -> InstructionResult {
```
Comment (DaniPopes) [inline crates/cheatcodes/src/inspector.rs:237]: This result is not used

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/cheatcodes/src/evm.rs (hunk 7 above)
Comment (ditto-eth) [reply crates/cheatcodes/src/evm.rs:13]: I think there was a type conflict before all the migrations maybe? but i changed it and it works fine, good catch!

File: crates/cheatcodes/src/evm.rs (hunk 8 above)
Comment (ditto-eth) [reply crates/cheatcodes/src/evm.rs:310]: possibly? I wasn't sure how to do this, since i want to track the address state separately from the address's slots? I could make another struct that contains the state + the map?

File: crates/cheatcodes/src/inspector.rs (hunk 9 above)
Comment (ditto-eth) [reply crates/cheatcodes/src/inspector.rs:239]: right now it just checks `if state.addresses.get(&contract_address).is_some() {` to skip any further code per address, you mean it's probably cheaper to check map size?

File: crates/cheatcodes/src/inspector.rs (hunk 10 above)
Comment (ditto-eth) [reply crates/cheatcodes/src/inspector.rs:237]: looks like I added this because I copied the use of `try_or_continue!(interpreter.stack().peek(0))` which returns `InstructionResult::Continue` on Err, but then step_end has an eval that is InstructionResult so wasn't sure if that is supposed to be returned instead? Otherwise I'd just inline the function back to make it work

File: crates/cheatcodes/src/evm.rs (hunk 8 above)
Comment (DaniPopes) [reply crates/cheatcodes/src/evm.rs:310]: you can use a tuple for `(AddressState, HashMap...)`

File: crates/cheatcodes/src/inspector.rs (hunk 9 above)
Comment (DaniPopes) [reply crates/cheatcodes/src/inspector.rs:239]: Yes, it is

File: crates/cheatcodes/src/evm.rs (hunk 8 above)
Comment (ditto-eth) [reply crates/cheatcodes/src/evm.rs:310]: oh nice! is that better/easier than making a `pub struct CoolState { pub address: AddressState .. }`?

Comment (DaniPopes) [reply crates/cheatcodes/src/evm.rs]: It's more convenient since you're only using this struct in this one map and for one purpose

Comment (ditto-eth) [reply crates/cheatcodes/src/evm.rs]: ok makes sense, I just pushed with the struct changes but can convert it

File: crates/cheatcodes/src/inspector.rs (hunk 10 above)
Comment (DaniPopes) [reply crates/cheatcodes/src/inspector.rs:244]: Ah that's fine then

File: crates/cheatcodes/src/evm.rs (hunk 8 above)
Comment (ditto-eth) [reply crates/cheatcodes/src/evm.rs]: though `if cool_state.address == AddressState::Cool` reads better than `if cool_state.0 == AddressState::Cool` unless we're ok with that? i assume there's a way to name them too

Comment (DaniPopes) [reply crates/cheatcodes/src/evm.rs]: you can destructure tuples like anything else `let (a, b) = ...`, `if let Some((a, b)) = ...`

Comment (ditto-eth) [reply crates/cheatcodes/src/evm.rs]: cool, seems to work

**Comment by ditto-eth** [comment]
@emo-eth informed me that #6310 landed which does similar things it seems - should I refactor this again to somehow use that data/code? if so, I think I'll need a lot of help to continue

**Comment by mattsse** [comment]
hey @ditto-eth 

> which does similar things it seems

Can you elaborate on whether there are synergies between the two?
I'm a bit out of the loop here but happy to help

sorry for the late reply

**Comment by zerosnacks** [comment]
Optimistically closing, I feel like this case is sufficiently covered by `--isolate` and our gas snapshot cheatcodes. If you feel strongly about this we can re-open / reconsider. Thanks for your PR and your effort!
